shadow/pkg/handler/deployment_cloner.go

```diff
@@ -0,0 +1,251 @@
+/*
+ * Copyright (c) 2017, MegaEase
+ * All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package handler
+
+import (
+	"encoding/json"
+	"reflect"
+
+	"github.com/megaease/easemesh/mesh-shadow/pkg/object"
+	installbase "github.com/megaease/easemeshctl/cmd/client/command/meshinstall/base"
+	"github.com/pkg/errors"
+	appsV1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ShadowDeploymentFunc type ShadowFunc func(ctx *object.CloneContext) error
+type ShadowDeploymentFunc func() error
+
+type cloneDeploymentSpecFunc func(sourceDeployment *appsV1.Deployment, shadowService *object.ShadowService) *appsV1.Deployment
+
+func (cloner *ShadowServiceCloner) cloneDeployment(sourceDeployment *appsV1.Deployment, shadowService *object.ShadowService) ShadowDeploymentFunc {
+	shadowDeployment := cloner.cloneDeploymentSpec(sourceDeployment, shadowService)
+	return func() error {
+		err := installbase.DeployDeployment(shadowDeployment, cloner.KubeClient, shadowDeployment.Namespace)
+		if err != nil {
+			return errors.Wrapf(err, "Clone deployment %s for service %s failed", sourceDeployment.Name, shadowService.ServiceName)
+		}
+		return err
+	}
+}
+
+func (cloner *ShadowServiceCloner) cloneDeploymentSpec(sourceDeployment *appsV1.Deployment, shadowService *object.ShadowService) *appsV1.Deployment {
+	shadowDeployment := cloner.injectShadowConfiguration(
+		cloner.shadowDeploymentBaseSpec(
+			cloner.shadowDeploymentInitialize(nil)))(sourceDeployment, shadowService)
```

Comment:
the long code line is hard to read, propose to split to short statements.
and it seems we don't need to use closure functions, they make code hard to read.